### DISCUSSION TITLE
Output file

### DISCUSSION THREAD

**Discussion by MathXCruz** [microcombustion/ctwrap#84] [Troubleshooting] [answered]
Hello,

I tried to run an example case, like

 ctwrap run freeflame freeflame.yaml --parallel --strategy sequence

and I got some warnings, with this error message:

RuntimeWarning: Output of metadata failed with error message:
Unable to open file (unable to open file: name = 'freeflame.h5', errno = 2, error message = 'No such file or directory', flags = 1, o_flags = 2)

Do I need to create an empty freeflame.h5 file beforehand? If so, where should it be created? Also, can i choose csv instead of h5 in the YAML configuration file?

**Comment by ischoegl** [comment]
Shouldn't be necessary. Could you let me know what operating system this is running on? (also, what python version)

**Reply by MathXCruz** [reply]
I'm using Ubuntu 20.04 and Python 3.8.5.

**Reply by ischoegl** [reply]
Are you running in a conda environment or did you install using `pip`?

**Reply by MathXCruz** [reply]
I'm running in a conda environment.

**Reply by ischoegl** [reply]
Quite odd - I'm doing the same here, albeit on Fedora with a python version 3.7.10. Did you create the environment with the `environment.yml` file as discussed [here](https://microcombustion.github.io/ctwrap/installation.html)?

**Reply by MathXCruz** [reply]
I just uninstalled and reinstalled it to make sure I didn't miss any step and I still get the same error. I probably did something wrong before though, because now it's using the same python version as you are. I believe the first time I installed it, I used pip to install it inside my cantera environment, now I created the ctwrap environment following the Full Instalation procedure.

**Reply by ischoegl** [reply]
could you run `pip uninstall ctwrap` outside of the conda environment? Also, what output do you get for
```
which ctwrap
```
Please also post the entire error stack as some of the warnings may give clues.

**Reply by MathXCruz** [reply]
I don't have pip installed outside conda, so it doesn't work.

 Running _which ctwrap_, I get

 /home/matheus/anaconda3/envs/ctwrap/bin/ctwrap

 and the output from running, with the extra -v, is below.

################################################################################
Running simulations: module `ctwrap.modules.freeflame`
Conditions specified in `freeflame.yaml` 
################################################################################
Simulations for entry `upstream.phi` with values: [0.4, 0.6, 0.8, 1.0, 1.2, 1.4,
1.6, 1.8, 2.0, 2.2, 2.4, 2.6]
 * Starting serial batch simulation
 * running `case_00` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_00' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_01` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_01' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_02` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_02' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_03` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_03' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_04` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_04' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_05` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_05' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_06` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_06' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_07` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_07' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_08` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_08' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_09` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_09' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_10` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_10' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
 * running `case_11` (main)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/handler.py:522: RuntimeWarning: Simulation of 'Freeflame' for 'case_11' failed with error message:
Cannot construct 'Parser' object from freeflame.yaml with type '<class 'str'>'
  warnings.warn(msg, RuntimeWarning)
/home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/ctwrap/output.py:349: RuntimeWarning: Output of metadata failed with error message:
Unable to open file (unable to open file: name = 'freeflame.h5', errno = 2, error message = 'No such file or directory', flags = 1, o_flags = 2)
  warnings.warn(msg, RuntimeWarning)

**Reply by ischoegl** [reply]
Alright. I think we're getting closer as it looks like there's something amiss with the parser. Could you run
```
pytest tests/test_parser.py
```
from the root folder? (this is the test suite for the parser)

If it runs without errors, also try `pytest` for everything.

**Reply by MathXCruz** [reply]
I ran it an this is the output.
After that, I tried searching for the tests folder and found out there is another ctwrap folder which is in my home folder and contains the tests folder, is that OK?

==================================== test session starts =====================================
platform linux -- Python 3.7.10, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
rootdir: /
plugins: cov-2.11.1
collected 0 items                                                                            

====================================== warnings summary ======================================
home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/_pytest/cacheprovider.py:428
  /home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/_pytest/cacheprovider.py:428: PytestCacheWarning: could not create cache path /.pytest_cache/v/cache/nodeids
    config.cache.set("cache/nodeids", sorted(self.cached_nodeids))

home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/_pytest/stepwise.py:49
  /home/matheus/anaconda3/envs/ctwrap/lib/python3.7/site-packages/_pytest/stepwise.py:49: PytestCacheWarning: could not create cache path /.pytest_cache/v/cache/stepwise
    session.config.cache.set(STEPWISE_CACHE_DIR, [])

-- Docs: https://docs.pytest.org/en/stable/warnings.html
==================================== 2 warnings in 0.01s =====================================
ERROR: file or directory not found: tests/test_parser.py

**Reply by ischoegl** [reply]
Just to make sure: please run it from the same folder where you ran the setup for conda etc.. The conda environment also has to be activated for this to work. I’ll probably push an update that allows to troubleshoot it further over the weekend

**Reply by MathXCruz** [reply]
When running it from the installation folder I get

ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov=ctwrap
  inifile: /home/matheus/ctwrap/pytest.ini
  rootdir: /home/matheus/ctwrap

**Comment by ischoegl** [comment]
Per your other comment: CSV output currently doesn't appear to work, but that's likely not a big issue to fix (see #86). The advantage of H5 is that you can re-import a solution into a cantera FreeFlame object at a later point.

**Comment by ischoegl** [comment]
@MathXCruz ... one thing I just remembered is that `MultiProcessing` really messes with error outputs. Could you run the same job sequentially (non-parallel):
```
ctwrap run freeflame freeflame.yaml --strategy sequence -v
```
(I am throwing the 'v' flag in here so you see more output)

**Comment by ischoegl** [comment]
@MathXCruz ... I pushed an update to the repository that should help diagnose the issue (I believe there is something not quite right about the conda environment). From the installation folder, run

```
git pull
python setup.py develop
```
to update to the newest version, and then run
```
ctwrap test freeflame
```
which should give the actual error stack. (I added this option as the package converts errors to warnings for batch jobs, which loses important information that I need to diagnose the issue).

**Reply by MathXCruz** [reply]
Here is the message I get by following these steps.

usage: ctwrap [-h] {list,run} ...
ctwrap: error: argument command: invalid choice: 'test' (choose from 'list', 'run')

**Reply by ischoegl** [reply]
Looks like it didn’t update. Could you run `pip uninstall ctwrap` followed by `pip install . -e`?

**Reply by MathXCruz** [reply]
`pip install . -e` doesn't work, I tried `pip install . -e ctwrap` and the result is:

ERROR: File "setup.py" not found. Directory cannot be installed in editable mode: /home/matheus/ctwrap/ctwrap

The weird thing is I can see setup.py in the ctwrap folder.

**Reply by ischoegl** [reply]
Ok. Did you activate the conda environment?

**Reply by MathXCruz** [reply]
I did. I cannot use pip without activating it.

**Comment by ischoegl** [comment, marked as answer]
Alright. I believe there's something awry with the conda environment. Could you start over, and make sure that the old environment is completely removed? I.e. start with a deactivated environment in the installation folder (below, I am assuming that your conda environment is 'ctwrap'). Start by removing the old version
```
conda remove --name ctwrap --all
```
and then reinstall as follows:
```
conda env create -f environment.yml
conda activate ctwrap
git pull
pip install -e .
```
After this, test it by running
```
ctwrap test freeflame
```

**Reply by MathXCruz** [reply]
It works! I also tested

`ctwrap run freeflame freeflame.yaml --parallel --strategy sequence`

and it works just fine. Thank you very much!

**Reply by ischoegl** [reply]
There may have been some remnants from a previous conda environment. Really happy that this is resolved!
